### ISSUE TITLE
OpenBLAS broken on Ubuntu 20.04

### ISSUE THREAD

**Issue by conradsnicta** [xianyi/OpenBLAS#2642]
This is a heads up in case other people run into this problem.

It appears that OpenBLAS 0.3.8 shipped with Ubuntu 20.04 is broken. Many packages relying on OpenBLAS will fail or freeze.  This includes Octave, mlpack, Armadillo, etc.  The main problem seems to be compilation flags specific to Ubuntu.

Partial workaround (until a proper fix for Ubuntu 20.04 is released):
1. Remove all default `libopenblas` packages, such as `libopenblas-dev`, `libopenblas-pthread-dev`,  `libopenblas0`, `libopenblas0-serial`, `libopenblas0-pthread`, etc.
2. Install only the `libopenblas-openmp-dev` package.

Related bug reports:
* https://bugs.launchpad.net/ubuntu/+source/openblas/+bug/1860601
* https://bugs.launchpad.net/ubuntu/+source/openblas/+bug/1870138
* https://bugs.debian.org/961725
* https://github.com/mlpack/mlpack/issues/2437

**Comment by martin-frbg** [comment]
There is also https://github.com/Reference-LAPACK/lapack/issues/407
Would it make sense to try and override any `-Wl,-Bsymbolic_functions` seen in LDFLAGS at the makefile level ?

**Comment by conradsnicta** [comment]
Rather than overriding flags, perhaps print a big fat warning, or better yet, abort compilation ?

Overriding flags and continuing compilation would in effect imply that OpenBLAS is responsible for fixing other projects' mistakes.  That's a slippery slope and sets up expectations.

`-Wl,-Bsymbolic_functions` also looks very gcc specific. Other compilers (such as clang) may have equivalent/similar functionality that's enabled with different flags, so this may end up being a whack-a-mole situation.

**Comment by martin-frbg** [comment]
Right now I am not even sure I understand exactly why this option causes breakage,although
https://software.intel.com/content/www/us/en/develop/articles/performance-tools-for-software-developers-bsymbolic-can-cause-dangerous-side-effects.html hints at some "nice" possibilities

**Comment by martin-frbg** [comment]
Found a couple of old blog posts on the general topic here - https://flameeyes.blog/tag/collisions/ , and his tool to find "link collisions" (which appears to be what the `-Bsymbolic-functions` is supposed to work around) is available as part of a ruby package https://github.com/Flameeyes/ruby-elf/tree/master/tools/link-collisions 
(Still unclear which function name exposed by OpenBLAS could collide with either libc or pthreads, plain `nm -a` does not bring up anything immediately suspicious. If I assume that it has to be one of the symbols flagged as "weak" by nm, about the only likely candidate is omp_in_parallel)

**Comment by martin-frbg** [comment]
Also https://bugs.launchpad.net/ubuntu/+source/libxfont/+bug/230460 (and their bug 226156 linked from it) seems to have highlighted the general problem with having -Bsymbolic_functions as the default buildpackage-LDFLAGS **in 2008**
and a comment in https://bugs.launchpad.net/ubuntu/+source/libxfont/+bug/226156 explains
(quoting a chat apparently)
```
<slangasek> bryce: -Wl,-Bsymbolic-functions is a single option; its purpose is to optimize the start-time symbol resolution at the expense of some "correctness", by causing any references to symbols available within the lib itself to be bound at build time
<bryce> slangasek: ah; is it a new addition? is there risk in turning it off?
<slangasek> bryce: there's no risk at all in turning it off, AFAIK it's entirely a performance thing
```
Yet they still seem to be doing it by default, only disabling on a case-by-case basis (which probably goes wrong sometimes as upstream code, maintainers or build scripts change).

**Comment by martin-frbg** [comment]
Comment on the corresponding LAPACK bug confirms that the Ubuntu20.04 problem hit the netlib reference implementation as well (i.e. the affected user did not use OpenBLAS through the /etc/alternatives mechanism), so this cannot be related to any recent thread-safety fixes. So closing here as the Ubuntu maintainers have made updated packages available.

**Comment by conradsnicta** [comment]
A proposed fix is available. However, it won't be pushed as an official update in Ubuntu 20.04 until people actually report that it fixes their problem. It's an overly conservative and cumbersome approach, but there you go.

More details: https://bugs.launchpad.net/ubuntu/+source/openblas/+bug/1860601/comments/9

**Comment by conradsnicta** [comment]
A proposed fix is available.  However, it won't be pushed as an official update in Ubuntu 20.04 until people actually report that it fixes their problem.  This can be done by replying to the Ubuntu bug report on Launchpad.

See comment 9 in https://bugs.launchpad.net/ubuntu/+source/openblas/+bug/1860601

**Comment by conradsnicta** [comment]
Fix finally released for Ubuntu 20.04.

    $ sudo aptitude show libopenblas0
    Package: libopenblas0                    
    Version: 0.3.8+ds-1ubuntu0.20.04.1

More details at https://bugs.launchpad.net/ubuntu/+source/openblas/+bug/1860601
